crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1362,37 +1416,45 @@ impl Loader<ParentVersionKey> for Db {
                 })
             });
 
-        // Wait for the reads to all finish, and gather them into the result map.
         let groups = futures::future::join_all(futures).await;
-
-        let mut results = HashMap::new();
+        let mut group_map = HashMap::new();
+        let mut historical_keys = vec![];
         for group in groups {
             for (group_key, stored) in
                 group.map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?
             {
                 // This particular object is invalid -- it didn't exist at the checkpoint we are
                 // viewing at.
-                if group_key.checkpoint_viewed_at < stored.checkpoint_sequence_number as u64 {
+                if group_key.checkpoint_viewed_at < stored.cp_sequence_number as u64 {
                     continue;
                 }
-
-                let object = Object::try_from_stored_history_object(
-                    stored,
-                    group_key.checkpoint_viewed_at,
-                    // If `LatestAtKey::parent_version` is set, it must have been correctly
-                    // propagated from the `Object::root_version` of some object.
-                    Some(group_key.parent_version),
-                )?;
+                let id = addr(&stored.object_id)?;
+                let historical_key = HistoricalKey {
+                    id,
+                    version: stored.object_version as u64,
+                    checkpoint_viewed_at: group_key.checkpoint_viewed_at,
+                };
+                historical_keys.push(historical_key);
 
                 let key = ParentVersionKey {
-                    id: object.address,
+                    id,
                     checkpoint_viewed_at: group_key.checkpoint_viewed_at,
                     parent_version: group_key.parent_version,
                 };
-
-                results.insert(key, object);
+                group_map.entry(key).or_insert(vec![]).push(historical_key);
```

Comment:
nit: avoid creating the default when we don't need it (although in the case of an empty vector, the cost is actually minimal because it doesn't allocate until the first push).

```suggestion
                group_map.entry(key).or_default().push(historical_key);
```